core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java

```diff
@@ -176,7 +176,7 @@ public void execute(Runnable command) {
    @Override
    public <I, O> CompletionStage<O> thenApplyBlocking(CompletionStage<? extends I> stage,
          Function<? super I, ? extends O> function, Object traceId) {
-      if (isCurrentThreadBlocking()) {
+      if (CompletionStages.isCompletedSuccessfully(stage) && isCurrentThreadBlocking()) {
```

Comment:
BTW, wouldn't make more sense to check if it is completed, successfully or not, and handle exceptions?

Reply:
I had debated about that, but I personally don't feel it is really worth it to worry about exceptions for the invoking blocking thread path. If anything I would say we can change that in a separate PR if we really wanted.